src/oc/code.cpp

```diff
@@ -22,123 +19,36 @@
-    if (stackp >= stacklast) \
-        execerror("Stack too deep.", "Increase with -NSTACK stacksize option");
-
-int tstkchk_actual(int i, int j) {
-    int k, l;
-    char* s[2];
-    if (i != j) {
-        for (k = 0, l = i; k < 2; k++, l = j) {
-            switch (l) {
-            case NUMBER:
-                s[k] = "(double)";
-                break;
-            case STRING:
-                s[k] = "(char *)";
-                break;
-            case OBJECTVAR:
-                s[k] = "(Object **)";
-                break;
-            case USERINT:
-                s[k] = "(int)";
-                break;
-            case SYMBOL:
-                s[k] = "(Symbol)";
-                break;
-            case VAR:
-                s[k] = "(double *)";
-                break;
-            case OBJECTTMP: /* would use OBJECT if it existed */
-                s[k] = "(Object *)";
-                break;
-            case STKOBJ_UNREF: /* hoc_stkobj_unref allready called */
-                s[k] = "(Object * already unreffed on stack)";
-                break;
-            default:
-                s[k] = "(Unknown)";
-                break;
-            }
-        }
-        fprintf(stderr, "bad stack access: expecting %s; really %s\n", s[1], s[0]);
-        execerror("interpreter stack type error", (char*) 0);
-    }
-    return 0;
-}
-
-#define USEMACROS 1
-
-/* warning! tstkchk(i,j) when i!=j will call execerror and error recovery
-   now uses stackp to recover OBJECTTMP resources. So it must be the case that
-   stackp - stack is an even number (since stack item, itemtype values
-   use a pair of stack locations). This invalidates the previous pop idiom
-   tstkchk((--stackp)->i, type), (--stackp)->val)) since if tstkchk calls
-   execerror without returning, stackp is no longer consistent since the
-   second decrement no longer takes place.
-
-   Furthermore, tstkchk(i,j) should be called prior to actually popping the
-   stack so that the execerror will properly unref the otherwise unexpected
-   possible OBJECTTMP.
-*/
-
-#if USEMACROS
-/* warning! tstkchk is a macro that uses each arg twice. So error if
-   the arg in the call has side effects. Eg avoid args like --stackp
-*/
-#define tstkchk(i, j) (((i) != (j)) ? tstkchk_actual(i, j) : 0)
-#define pushxm(d)            \
-    ((stackp++)->val = (d)); \
-    ((stackp++)->i = NUMBER)
-#define pushsm(d)            \
-    ((stackp++)->sym = (d)); \
-    ((stackp++)->i = SYMBOL)
-#define nopopm() (stackp -= 2) /*provision at use made to deal with OBJECTTMP*/
-#define xpopm()  (tstkchk(stackp[-1].i, NUMBER), nopopm(), stackp->val)
-#define spopm()  (tstkchk(stackp[-1].i, SYMBOL), nopopm(), stackp->sym)
-#else
-#define pushxm(d)     pushx(d)
-#define pushsm(d)     pushs(d)
-#define xpopm()       xpop()
-#define spopm()       spop()
-#define nopopm()      nopop()
-#define tstkchk(i, j) tstkchk_actual(i, j)
-#endif
-
-#define EPS hoc_epsilon
 
-#define NSTACK 1000 /* default size */
```

Comment:
Now that there is no more NSTACK, it can be removed from the optarg of nrniv and the properties in ivocmain.cpp and nrn.defaults.in. And the docs?

Reply:
Actually it still exists "really":
https://github.com/neuronsimulator/nrn/blob/89c8c572952618557c3cadfee22172719f77477c/src/oc/code.cpp#L394

I didn't try to enable dynamically growing the stack side:
https://github.com/neuronsimulator/nrn/blob/89c8c572952618557c3cadfee22172719f77477c/src/oc/code.cpp#L730-L732

as I wasn't sure if this was safe, given that I understand we sometimes store pointers to values on the stack. That seemed a reasonable topic for follow up later...

Reply:
I believe the original logic has been kept. Could we get rid of it and rely on vector reallocation strategy? not sure that would work.

Reply:
The only difficulty I'm aware of is when there is a pointer to a stack element (Object** to an Object* on the stack). It may be the case that the only time Object* is on the stack is for ```localobj``` object variables. Also, I guess, the frames for nested calls are on the stack. ```struct frame``` has a field ```StackDatum* argn;      /* n-th argument on stack */``` but I suppose that could become an index.

Reply:
I think this can be deferred to a follow up task. Another option could be to use `std::deque` instead, as then growing the storage does not invalidate existing elements, and we would just need to ensure that we avoid any problematic pointer arithmetic.